GooglePlayInstant/Editor/QuickDeploy/LoadingScreenGenerator.cs

```diff
@@ -0,0 +1,159 @@
+            EditorSceneManager.CloseScene(SceneManager.GetSceneByName(Path.GetFileNameWithoutExtension(SceneName)),
+                true);
+
+            var loadingScreenScene = EditorSceneManager.NewScene(NewSceneSetup.EmptyScene, NewSceneMode.Additive);
+
+            var loadingScreenGameObject = new GameObject(CanvasName);
+
+            AddImageToScene(loadingScreenGameObject, loadingScreenImagePath);
+
+            AddScript(loadingScreenGameObject);
+
+            LoadingBar.AddComponent(loadingScreenGameObject);
+
+            bool saveOk = EditorSceneManager.SaveScene(loadingScreenScene, SceneFilePath);
+
+            if (!saveOk)
+            {
+                // Not a fatal issue. User can attempt to resave this scene.
+                var warningMessage = string.Format("Issue while saving scene {0}.",
+                    SceneName);
+
+                Debug.LogWarning(warningMessage);
+
+                DialogHelper.DisplayMessage(SaveErrorTitle, warningMessage);
+            }
+            else
+            {
+                //TODO: investigate GUI Layout errors that occur when moving this to DialogHelper
+                if (EditorUtility.DisplayDialog("Change Scenes in Build",
+                    "Would you like to replace any existing Scenes in Build with the loading screen scene?", "Yes",
+                    "No"))
+                {
+                    SetMainSceneInBuild(SceneFilePath);
+                }
+            }
+        }
+
+        public static bool LoadingScreenExists()
+        {
+            return SceneManager.GetSceneByName(Path.GetFileNameWithoutExtension(SceneName)).IsValid();
+        }
+
+        public static GameObject GetLoadingScreenCanvasObject()
+        {
+            return GameObject.Find(CanvasName);
+        }
+
+        // Visible for testing
+        internal static void SetMainSceneInBuild(string pathToScene)
+        {
+            EditorBuildSettings.scenes = new[]
+            {
+                new EditorBuildSettingsScene(pathToScene, true)
+            };
+        }
+
+        // Visible for testing
+        internal static void AddScript(GameObject loadingScreenGameObject)
+        {
+            loadingScreenGameObject.AddComponent<LoadingScreenScript>();
+        }
+
+
+        // Visible for testing
+        internal static void AddImageToScene(GameObject loadingScreenGameObject,
+            string pathToLoadingScreenImage)
+        {
+            if (loadingScreenGameObject.GetComponent<Canvas>() == null)
+            {
+                // First time creating a loading screen, configure nested game objects appropriately.
+                var loadingScreenCanvas = loadingScreenGameObject.AddComponent<Canvas>();
+
+                loadingScreenCanvas.renderMode = RenderMode.ScreenSpaceOverlay;
+
+                loadingScreenGameObject.AddComponent<Image>();
+            }
+
+            var loadingScreenImageData = File.ReadAllBytes(pathToLoadingScreenImage);
```

Comment:
Instead of reading the bytes directly from the image try using 'AssetDatabase.LoadAssetAtPath()'.
Reading the bytes directly and making a new sprite breaks Unity's reference to the Asset. If the user makes any changes to that file, or to its meta file (ex. changing compression settings), it won't be reflected in the loading scene.

Reply:
AssetDatabase can only load assets relative to the project folder. I opted to use File.ReadAllBytes so that the Open File panel thats associated with the pathToLoadingScreenImage can be retrieved from anywhere on disk currently. I thought the flow might be better that way, instead of limiting them to resources in their Project folder. 

Currently, the UI supports this, by sporting an "Update Loading Scene" button that would suggest one has to click this in order to sync new changes to assets. (instead of automatically changing based on the chosen loading screen image -- a field that's even currently not persisted d between sessions, or in other words saved in the Config file.)

Is this new behavior preferred though? I thought the current way would be better for the user potentially.

Reply:
That's a good point, and I think it comes down to how much the plugin "owns" that loading screen. Is the user allowed or expected to modify the loading scene themselves? Or is the expectation that they shouldn't touch the scene hierarchy, and only modify the scene through what we expose in that window?

If the plugin has ownership over that scene, then I think the current approach works. Although I would maybe check if the path is in the Assets folder, and do AssetDatabase.LoadAssetAtPath() if it is.

If the expectation is that the plugin generates a default loading scene, which the developer can then modify, I think we should reference the Texture2D directly instead of using a path. A good example of Unity using this method is in PlayerSettings -> Splash Image -> Static Splash image.